FALCON 1861 : Support HDFS Snapshot based replication in Falcon

bvellanki opened the pull request:
Documentation will be added in Jira FALCON-1908

sandeepSamudrala commented on addons/extensions/hdfs-snapshot-mirroring/src/main/META/hdfs-snapshot-mirroring-properties.json:
```diff
@@ -0,0 +1,173 @@
+      "example":"10"
+    },
+    {
+      "propertyName":"targetCluster",
```
jobClusterName and targetCluster should be same?

bvellanki replied:
Not necessary. Job can be run on either source or target

sandeepSamudrala commented on addons/extensions/hdfs-snapshot-mirroring/src/main/resources/runtime/hdfs-snapshot-mirroring-secure-workflow.xml:
```diff
@@ -0,0 +1,179 @@
+            <arg>-targetNNKerberosPrincipal</arg>
+            <arg>${targetNNKerberosPrincipal}</arg>
+            <arg>-targetSnapshotDir</arg>
+            <arg>${targetSnapshotDir}</arg>
```
Take directory paths from wf conf " ${wf:conf(" as in case of a directory having special characters like ".,-" might cause issues.

bvellanki replied:
This has to be done for all extensions. I will handle this is a separate Jira.

sandeepSamudrala commented on addons/extensions/hdfs-snapshot-mirroring/src/main/resources/runtime/hdfs-snapshot-mirroring-secure-workflow.xml:
```diff
@@ -0,0 +1,179 @@
+            <arg>-targetNNKerberosPrincipal</arg>
+            <arg>${targetNNKerberosPrincipal}</arg>
+            <arg>-targetSnapshotDir</arg>
+            <arg>${targetSnapshotDir}</arg>
```
Take directory paths from wf conf " ${wf:conf(" as in case of a directory having special characters like ".,-" might cause issues.
Can you make changes in all locations where directory is being used?

bvellanki replied:
Same as above

sandeepSamudrala commented on addons/extensions/hdfs-snapshot-mirroring/src/main/resources/runtime/hdfs-snapshot-mirroring-secure-workflow.xml:
```diff
@@ -0,0 +1,179 @@
+        <ok to="snapshot-retention"/>
+        <error to="fail"/>
+    </action>
+    <!-- Export Replication action -->
```
Correct the comment. Ths action is retention while comment says replicaiton action

bvellanki replied:
Will fix this. Thanks @sandeepSamudrala

sandeepSamudrala commented on addons/extensions/hdfs-snapshot-mirroring/src/main/resources/runtime/hdfs-snapshot-mirroring-workflow.xml:
```diff
@@ -0,0 +1,165 @@
+        <ok to="snapshot-retention"/>
+        <error to="fail"/>
+    </action>
+    <!-- Export Replication action -->
```
minor nit: update the comment over here too..

bvellanki replied:
Will fix this

sandeepSamudrala commented on addons/hdfs-snapshot-mirroring/README:
```diff
@@ -0,0 +1,30 @@
+HDFS Snapshot Mirroring Extension.
+
+Overview
+This extension implements replication for snapshot-able directories on HDFS from one
```
replication along with retention?

sowmyaramesh commented on addons/extensions/hdfs-snapshot-mirroring/pom.xml:
```diff
@@ -0,0 +1,32 @@
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>org.apache.falcon.recipes</groupId>
```
change recipes to extensions. Did it generate the jar with this issue?

bvellanki replied:
I will change this. It did generate the jar.

sowmyaramesh commented on addons/extensions/hdfs-snapshot-mirroring/src/main/resources/runtime/hdfs-snapshot-mirroring-secure-workflow.xml:
```diff
@@ -0,0 +1,179 @@
+            <arg>${distcpMapBandwidth}</arg>
+            <arg>-sourceNN</arg>
+            <arg>${sourceNN}</arg>
+            <arg>-sourceExecUrl</arg>
```
What is sourceExecUrl for?

bvellanki replied:
sourceExecUrl (target and job as well) is added to Configuration needed to create DistributedFileSystem in HdfsSnapshotReplicator and HdfsSnapshotEvictor.

sowmyaramesh commented on addons/extensions/hdfs-snapshot-mirroring/src/main/resources/runtime/hdfs-snapshot-mirroring-secure-workflow.xml:
```diff
@@ -0,0 +1,179 @@
+            <arg>-sourceExecUrl</arg>
+            <arg>${sourceExecUrl}</arg>
+            <arg>-sourceNNKerberosPrincipal</arg>
+            <arg>${sourceNNKerberosPrincipal}</arg>
```
We should have only one WF for secure & unsecure and then add the additional properties at run time. This way we won't have multiple WF's which would avoid duplication, more readable less error prone in future if we have to modify the WF.

bvellanki replied:
Handled in separate Jira https://issues.apache.org/jira/browse/FALCON-1912 since this is applicable to all extensions.

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/replication/snapshots/HdfsSnapshotReplicator.java:
```diff
@@ -0,0 +1,305 @@
+            DistCpOptions options = getDistCpOptions(sourceStorageUrl, targetStorageUrl,
+                    sourceFs, targetFs, sourceDir, targetDir, currentSnapshotName);
+            DistCp distCp = new DistCp(jobConf, options);
+            LOG.info("Started Snapshot based DistCp from " + getStagingUri(sourceStorageUrl, sourceDir)
```
Please use {} instead of +

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/replication/snapshots/HdfsSnapshotReplicator.java:
```diff
@@ -0,0 +1,305 @@
+        DistributedFileSystem sourceFs = HadoopClientFactory.get().createDistributedProxiedFileSystem(sourceConf);
+        DistributedFileSystem targetFs = HadoopClientFactory.get().createDistributedProxiedFileSystem(targetConf);
+
+        String currentSnapshotName = "falcon-snapshot-"
```
Minot nit: create constant

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/replication/snapshots/HdfsSnapshotReplicator.java:
```diff
@@ -0,0 +1,305 @@
+
+        String currentSnapshotName = "falcon-snapshot-"
+                + cmd.getOptionValue(HdfsSnapshotMirrorProperties.SNAPSHOT_JOB_NAME.getName())
+                + "-" + System.currentTimeMillis();
```
minor nit: create constant

bvellanki replied:
"-" is only used at one location in the code. I dont think it makes sense to create a constant for this.

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/replication/snapshots/HdfsSnapshotReplicator.java:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    private String getStagingUri(String storageUrl, String dir) {
+        return storageUrl + dir;
```
Can we check if storageUrl ends with "/" else add it?

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/replication/snapshots/HdfsSnapshotReplicator.java:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    private String getSnapshotDir(String dirName) {
+        return dirName + Path.SEPARATOR + ".snapshot" + Path.SEPARATOR;
```
Minor nit:use constant for ".snapshot"

sowmyaramesh replied:
Can dirName have "/" at end? If so remove using StringUtils.removeEnd

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/retention/snapshots/HdfsSnapshotEvictor.java:
```diff
@@ -0,0 +1,228 @@
+                HdfsSnapshotMirrorProperties.SOURCE_SNAPSHOT_RETENTION_AGE_LIMIT.getName());
+        int numSnapshots = Integer.parseInt(
+                cmd.getOptionValue(HdfsSnapshotMirrorProperties.SOURCE_SNAPSHOT_RETENTION_NUMBER.getName()));
+        if (retPolicy.equalsIgnoreCase("delete")) {
```
MInor nit: use constant

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/retention/snapshots/HdfsSnapshotEvictor.java:
```diff
@@ -0,0 +1,228 @@
+                HdfsSnapshotMirrorProperties.TARGET_SNAPSHOT_RETENTION_AGE_LIMIT.getName());
+        numSnapshots = Integer.parseInt(
+                cmd.getOptionValue(HdfsSnapshotMirrorProperties.TARGET_SNAPSHOT_RETENTION_NUMBER.getName()));
+        if (retPolicy.equalsIgnoreCase("delete")) {
```
Minor nit: Use constant

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/retention/snapshots/HdfsSnapshotEvictor.java:
```diff
@@ -0,0 +1,228 @@
+
+            long evictionTime = System.currentTimeMillis() - EvictionHelper.evalExpressionToMilliSeconds(ageLimit);
+
+            String snapshotDir = dirName + Path.SEPARATOR + ".snapshot" + Path.SEPARATOR;
```
Minor nit: Use constant

sowmyaramesh replied:
if dirName can have "/" at end remove using StringUtils.removeEnd

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/retention/snapshots/HdfsSnapshotEvictor.java:
```diff
@@ -0,0 +1,228 @@
+
+    }
+
+    private CommandLine getCommand(String[] args) throws org.apache.commons.cli.ParseException {
```
I think few options are same for eviction and replication. Can we move it to util to reuse code?

sowmyaramesh commented on extensions/src/main/java/org/apache/falcon/extensions/mirroring/hdfsSnapshot/HdfsSnapshotMirrorProperties.java:
```diff
@@ -0,0 +1,90 @@
+
+    TDE_ENCRYPTION_ENABLED("tdeEncryptionEnabled", "Is TDE encryption enabled on source and target", false),
+
+    SNAPSHOT_JOB_NAME("snapshotJobName", "Name of cluster on which job will be run", false),
```
is this Name of cluster on which job will be run or typo?

sowmyaramesh commented on extensions/src/main/java/org/apache/falcon/extensions/mirroring/hdfsSnapshot/HdfsSnapshotMirrorProperties.java:
```diff
@@ -0,0 +1,90 @@
+    TARGET_NN_KERBEROS_PRINCIPAL("targetNNKerberosPrincipal",
+            "Snapshot replication target kerberos principal", false),
+
+    TARGET_SNAPSHOT_DIR("targetSnapshotDir", "Target Hive metastore uri", false),
```
ISn't this mandatory?

sowmyaramesh commented on src/main/assemblies/distributed-package.xml:
```diff
@@ -150,6 +150,40 @@
+            <directory>addons/extensions/hdfs-snapshot-mirroring/src/main/resources</directory>
+            <outputDirectory>extensions/hdfs-snapshot-mirroring/resources</outputDirectory>
+            <fileMode>0755</fileMode>
+            <directoryMode>0755</directoryMode>
```
You missed adding README in all the packaging xml's

sowmyaramesh commented on common/src/main/java/org/apache/falcon/entity/ClusterHelper.java:
```diff
@@ -71,6 +73,18 @@ public static Configuration getConfiguration(Cluster cluster) {
+        conf.set(HadoopClientFactory.FS_DEFAULT_NAME_KEY, storageUrl);
+        conf.set(HadoopClientFactory.MR_JT_ADDRESS_KEY, executeEndPoint);
+        conf.set(HadoopClientFactory.YARN_RM_ADDRESS_KEY, executeEndPoint);
+        if (StringUtils.isNotBlank(kerberosPrincipal)) {
```
use isSecurityEnabled instead?

bvellanki replied:
We cannot assume that the falcon server and the cluster have same security level. Falcon server can be unsecure, and cluster can be secure.

sowmyaramesh commented on extensions/src/main/java/org/apache/falcon/extensions/mirroring/hdfsSnapshot/HdfsSnapshotMirroringExtension.java:
```diff
@@ -0,0 +1,239 @@
+        additionalProperties.put(HdfsSnapshotMirrorProperties.SOURCE_EXEC_URL.getName(),
+                ClusterHelper.getMREndPoint(sourceCluster));
+        String sourceKerberosPrincipal = ClusterHelper.getPropertyValue(sourceCluster, SecurityUtil.NN_PRINCIPAL);
+        if (StringUtils.isNotBlank(sourceKerberosPrincipal)) {
```
Please get all the security related properties if security is enabled instead of checking for null

bvellanki replied:
We cannot assume that the falcon server and the cluster have same security level. Falcon server can be unsecure, and cluster can be secure.

sowmyaramesh commented on extensions/src/main/java/org/apache/falcon/extensions/mirroring/hdfsSnapshot/HdfsSnapshotMirroringExtension.java:
```diff
@@ -0,0 +1,239 @@
+        additionalProperties.put(HdfsSnapshotMirrorProperties.TARGET_EXEC_URL.getName(),
+                ClusterHelper.getMREndPoint(targetCluster));
+        String targetKerberosPrincipal = ClusterHelper.getPropertyValue(targetCluster, SecurityUtil.NN_PRINCIPAL);
+        if (StringUtils.isNotBlank(targetKerberosPrincipal)) {
```
Please get all the security related properties if security is enabled instead of checking for null

bvellanki replied:
We cannot assume that the falcon server and the cluster have same security level. Falcon server can be unsecure, and cluster can be secure.

sandeepSamudrala replied:
I agree with you @bvellanki . One more thing I wanted to know was, how about in case when copying from secured cluster to unsecured and vice-versa, will this extension work in similar fashion in both the cases(Secured cluster to normal cluster works)?

bvellanki replied:
In theory it should work. But I have not tested this end2end. I tested secure to secure, unsecure to unsecure.

vrangan replied:
There are issues with respect to configuration (with ipc fallback to simple auth enabled) and there are situations where the generated workflows should work independent of the cluster where they run is secure or not.   OOZIE-2356 allows the ability to skip credentials at action level depending on the cluster is secure or not which will be needed to fully realize this

sowmyaramesh commented on addons/extensions/hdfs-snapshot-mirroring/pom.xml:
```diff
@@ -0,0 +1,32 @@
+  limitations under the License.
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
Why are there two README's? There should be only one README(without.txt) under addons/extensions/hdfs-snapshot-mirroring/README.

Also you have addons/hdfs-snapshot-mirroring and one other under addons/extensions. What is the extra one for?

bvellanki replied:
addons/hdfs-snapshot-mirroring contains code.  addons/extensions/hdfs-snapshot-mirroring contains files that are used by falcon server side extensions code to submit an extension job.

sowmyaramesh replied:
After applying your patch I didnt see any code in addons/hdfs-snapshot-mirroring. I thought you added code in replication module? /src/main/java/org/apache/falcon/replication/snapshots/HdfsSnapshotReplicator.java

bvellanki replied:
"git apply" from command line will have code copied to addons/hdfs-snapshot-mirroring/src/main/java  and to addons/hdfs-snapshot-mirroring/src/test/java.

If you try applying patch using IntelliJ, it may not work as expected.

sowmyaramesh commented on extensions/src/main/java/org/apache/falcon/extensions/mirroring/hdfsSnapshot/HdfsSnapshotMirroringExtension.java:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    @Override
+    public void validate(final Properties extensionProperties) throws FalconException {
```
Can you help me understand how is the secure WF resource picked in case of secure setup?

bvellanki replied:
In case of secure setup, the hdfs-snapshot-mirroring-secure-workflow.xml should be installed by the extensions.

sowmyaramesh replied:
What do you mean by install? Extension artifacts are stored in HDFS and extension will only get teh corresponding resource based on name.
If we have diff WF for secure please you have to let teh extension known.

// Add -secure and update the resource name
            String resourceName = getName().toLowerCase() + SECURE_RESOURCE;
            additionalProperties.put(ExtensionProperties.RESOURCE_NAME.getName(), resourceName);

Please refer to HiveMirroringExtension

bvellanki replied:
I think it is better to remove two versions (secure and unsecure) of workflow and maintain a single version. I will fix accordingly.

sowmyaramesh replied:
I agree. But I thought you wanted to fix it with diff jira. If we want this current code to work with secure mode this fix is required.

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/retention/snapshots/HdfsSnapshotEvictor.java:
```diff
@@ -0,0 +1,228 @@
+
+            for (int i = 0; i < (snapshots.length - numSnapshots); i++) {
+                // delete if older than ageLimit while retaining numSnapshots
+                if (snapshots[i].getModificationTime() < evictionTime) {
```
<= ?

bvellanki replied:
< is correct, since we only want to delete if older than age

sowmyaramesh commented on extensions/src/main/java/org/apache/falcon/extensions/mirroring/hdfsSnapshot/HdfsSnapshotMirroringExtension.java:
```diff
@@ -0,0 +1,239 @@
+
+    }
+
+    private boolean isDirSnapshotable(DistributedFileSystem hdfs, Path path) throws FalconException {
```
Make the method static

bvellanki replied:
Why? can you please elaborate?

sowmyaramesh commented on extensions/src/main/java/org/apache/falcon/extensions/mirroring/hdfsSnapshot/HdfsSnapshotMirroringExtension.java:
```diff
@@ -0,0 +1,239 @@
+        return additionalProperties;
+    }
+
+    private void validateRetentionPolicy(String retentionPolicy) throws FalconException {
```
Make the method static

bvellanki replied:
Why? can you please elaborate?

sowmyaramesh replied:
Its a good programming practice. You can refer this http://stackoverflow.com/questions/2671496/java-when-to-use-static-methods

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/retention/snapshots/HdfsSnapshotEvictor.java:
```diff
@@ -0,0 +1,228 @@
+        return 0;
+    }
+
+    protected void evictSnapshots(DistributedFileSystem fs, String dirName, String ageLimit,
```
Can this be made static?

sowmyaramesh commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/replication/snapshots/HdfsSnapshotReplicator.java:
```diff
@@ -0,0 +1,305 @@
+        return 0;
+    }
+
+    private void createSnapshotInFileSystem(String dirName, String snapshotName,
```
Can you rename this to createSnapshot and make it static?

vrangan commented:
Thanks for working through this @bvellanki 

I went through the rest of the review comments.  There are some that we may not do (say '-' as a constant), but some of the other items like making a string a constant may be as simple as declaring them static final with a constant name (for example for .snapshot).

Also, logging with parameters is also consistent with rest of the code and is helpful.

When you make a new pull request you can fix some of these minor nits to enhance the code quality

bvellanki commented:
@vrangan @sandeepSamudrala @sowmyaramesh  : I incorporated the feedback into code and submitted updated patch. Please review

yzheng-hortonworks commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/snapshots/util/HdfsSnapshotUtil.java:
```diff
@@ -0,0 +1,67 @@
+        return HadoopClientFactory.get().createDistributedProxiedFileSystem(sourceConf);
+    }
+
+    public static DistributedFileSystem getTargetFileSystem(CommandLine cmd) throws FalconException {
```
Can combine getTargetFileSystem and getSourceFileSystem to one function, e.g. getFileSystem(url, endpoint, principal).

yzheng-hortonworks commented:
+1

bvellanki commented:
I found a bug in the way jobConf is set for DistCP in kerberos secured environment. Fixed that issue and updated the patch. Please review and provide feedback.

vrangan commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/snapshots/retention/HdfsSnapshotEvictor.java:
```diff
@@ -0,0 +1,208 @@
+        Option opt = new Option(HdfsSnapshotMirrorProperties.SOURCE_NN.getName(), true, "Source Cluster");
+        opt.setRequired(true);
+        options.addOption(opt);
+        opt = new Option(HdfsSnapshotMirrorProperties.SOURCE_EXEC_URL.getName(),
```
We don't need this - we should cleanup these options later unless we want to enhance the code to make use of this.

vrangan commented on addons/hdfs-snapshot-mirroring/src/main/java/org/apache/falcon/snapshots/retention/HdfsSnapshotEvictor.java:
```diff
@@ -0,0 +1,208 @@
+                true, "Target snapshot-able dir to replicate");
+        opt.setRequired(true);
+        options.addOption(opt);
+        opt = new Option(HdfsSnapshotMirrorProperties.TARGET_EXEC_URL.getName(),
```
We don't need this - we should cleanup these options later unless we want to enhance the code to make use of this.

vrangan commented:
+1 - a few cleanups identified can be done later in a follow on work

sowmyaramesh commented:
+1, LGTM. Thanks for your contribution!
